Comptime reflection compares all field values as strings.

### Describe the bug

As said in the title, compile time reflection compares all field values as strings.
Looking at the generated C code:
```c
if (!string__eq(_v_new.ButtonStyleParams, _v_default.ButtonStyleParams)) {
	old->ButtonStyleParams = _v_new.ButtonStyleParams;
}
```

### Reproduction Steps

```v
struct Abc {
	id int
	name string
	letter string
}

fn join[T](mut old T, new T) {
	$if T is $struct {
		default := T{}
		$for field in T.fields {
			if new.$(field.name) != default.$(field.name) {
				old.$(field.name) = new.$(field.name)
			}
		}
	}
}

fn main() {
	mut a := Abc{name: 'Peter', letter: 'a'}
	b := Abc{id: 1, letter: 'b'}
	join(mut a, b)

	assert a == Abc{name: 'Peter', id: 1, letter: 'b'}
}
```

### Expected Behavior

Output:
```console
Abc{
    id: 1
    name: 'Peter'
    letter: 'b'
}
```

### Current Behavior

```console
==================
/tmp/v_1000/zest_launcher.tmp.c:76124: warning: assignment from incompatible pointer type
/tmp/v_1000/zest_launcher.tmp.c:91043: error: cannot convert 'struct ui__ButtonStyleParams' to 'struct string'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

0.4.2 8f76a06

### Environment details (OS name and version, etc.)

```
V full version: V 0.4.2 8f76a06
OS: linux, Linux version 6.1.52-1-lts (linux-lts@archlinux) (gcc (GCC) 13.2.1 20230801, GNU ld (GNU Binutils) 2.41.0) #1 SMP PREEMPT_DYNAMIC Thu, 07 Sep 2023 05:17:41 +0000
Processor: 8 cpus, 64bit, little endian, 11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz

getwd: /home/katekyy/programming/zest_launcher
vexe: /home/katekyy/.local/share/v/v
vexe mtime: 2023-10-14 18:01:00

vroot: OK, value: /home/katekyy/.local/share/v
VMODULES: OK, value: /home/katekyy/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.42.0
Git vroot status: weekly.2023.33-352-g8f76a062
.git/config present: true

CC version: cc (GCC) 13.2.1 20230801
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3
```
> [!NOTE]
> You can vote for this issue using the 👍 reaction. More votes increase the issue's priority for developers.
>
> Take into account that only the 👍 reaction counts as a vote.
> Only reactions to the issue itself will be counted as votes, not comments.

--- a/vlib/v/gen/c/infix.v
+++ b/vlib/v/gen/c/infix.v
@@ -96,8 +96,18 @@ fn (mut g Gen) infix_expr_arrow_op(node ast.InfixExpr) {
 
 // infix_expr_eq_op generates code for `==` and `!=`
 fn (mut g Gen) infix_expr_eq_op(node ast.InfixExpr) {
-	left := g.unwrap(node.left_type)
-	right := g.unwrap(node.right_type)
+	left_type := if node.left is ast.ComptimeSelector {
+		g.get_comptime_var_type(node.left)
+	} else {
+		node.left_type
+	}
+	right_type := if node.right is ast.ComptimeSelector {
+		g.get_comptime_var_type(node.right)
+	} else {
+		node.right_type
+	}
+	left := g.unwrap(left_type)
+	right := g.unwrap(right_type)
 	mut has_defined_eq_operator := false
 	mut eq_operator_expects_ptr := false
 	if m := g.table.find_method(left.sym, '==') {
@@ -110,7 +120,7 @@ fn (mut g Gen) infix_expr_eq_op(node ast.InfixExpr) {
 		g.gen_plain_infix_expr(node)
 		return
 	}
-	is_none_check := node.left_type.has_flag(.option) && node.right is ast.None
+	is_none_check := left_type.has_flag(.option) && node.right is ast.None
 	if is_none_check {
 		g.gen_is_none_check(node)
 	} else if (left.typ.is_ptr() && right.typ.is_int())

--- a/vlib/v/tests/comptime_eq_test.v
+++ b/vlib/v/tests/comptime_eq_test.v
@@ -1,0 +1,34 @@
+struct Abc {
+	id     int
+	name   string
+	letter string
+}
+
+fn join[T](mut old T, new T) {
+	$if T is $struct {
+		default := T{}
+		$for field in T.fields {
+			if new.$(field.name) != default.$(field.name) {
+				old.$(field.name) = new.$(field.name)
+			}
+		}
+	}
+}
+
+fn test_main() {
+	mut a := Abc{
+		name: 'Peter'
+		letter: 'a'
+	}
+	b := Abc{
+		id: 1
+		letter: 'b'
+	}
+	join(mut a, b)
+
+	assert a == Abc{
+		name: 'Peter'
+		id: 1
+		letter: 'b'
+	}
+}

